File: src/main/java/com/kelsos/mbrc/services/ProtocolHandler.java
package com.kelsos.mbrc.services;

import android.util.Log;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.kelsos.mbrc.BuildConfig;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.ProtocolEvent;
import com.kelsos.mbrc.others.Protocol;
import com.squareup.otto.Bus;
import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;

@Singleton
public class ProtocolHandler {
    public static double ServerProtocolVersion;
    private Bus bus;
    private boolean isHandshakeComplete;
    private ObjectMapper mapper;

    @Inject public ProtocolHandler(Bus bus, ObjectMapper mapper) {
        this.bus = bus;
        this.mapper = mapper;
    }

    public void resetHandshake() {
        isHandshakeComplete = false;
    }

    public void preProcessIncoming(final String incoming) {
        try {
            final String[] replies = incoming.split("\r\n");
            for (String reply : replies) {

                JsonNode node = mapper.readValue(reply, JsonNode.class);
                String context = node.path("context").getTextValue();

                if (context.contains(Protocol.ClientNotAllowed)) {
                    bus.post(new MessageEvent(ProtocolEvent.InformClientNotAllowed));
                    return;
                }

                if (!isHandshakeComplete) {
                    if (context.contains(Protocol.Player)) {
                        bus.post(new MessageEvent(ProtocolEvent.InitiateProtocolRequest));
                    } else if (context.contains(Protocol.Protocol)) {
                        ServerProtocolVersion = node.path("data").getDoubleValue();
                        isHandshakeComplete = true;
                        bus.post(new MessageEvent(ProtocolEvent.HandshakeComplete, true));
                    } else {
                        return;
                    }
                }

                bus.post(new MessageEvent(context, node.path("data")));
            }

        } catch (IOException e) {
            if (BuildConfig.DEBUG) {
                Log.d("mbrc-log", "Incoming message pre-processor", e);

            }
        }

    }
}

File: app/src/main/java/com/kelsos/mbrc/services/ProtocolHandler.kt
package com.kelsos.mbrc.services

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.kelsos.mbrc.constants.Const
import com.kelsos.mbrc.constants.Protocol
import com.kelsos.mbrc.constants.ProtocolEventType
import com.kelsos.mbrc.events.MessageEvent
import com.kelsos.mbrc.events.bus.RxBus
import com.kelsos.mbrc.model.MainDataModel
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProtocolHandler
  @Inject
  constructor(
    private val bus: RxBus,
    private val mapper: ObjectMapper,
    private val model: MainDataModel,
  ) {
    private var isHandshakeComplete: Boolean = false

    fun resetHandshake() {
      isHandshakeComplete = false
    }

    fun preProcessIncoming(incoming: String) {
      try {
        val replies =
          incoming
            .split("\r\n".toRegex())
            .dropLastWhile(String::isEmpty)
            .toTypedArray()

        replies.forEach {
          Timber.v("received:: $it")

          val node = mapper.readValue(it, JsonNode::class.java)
          val context =
            node
              .path("context")
              .textValue()
              .trim()
              .lowercase(Locale.getDefault())

          if (context == Protocol.ClientNotAllowed) {
            bus.post(MessageEvent(ProtocolEventType.InformClientNotAllowed))
            return
          }

          if (!isHandshakeComplete) {
            when (context) {
              Protocol.Player -> bus.post(MessageEvent(ProtocolEventType.InitiateProtocolRequest))
              Protocol.ProtocolTag -> handleProtocolMessage(node)
              else -> return
            }
          }

          bus.post(MessageEvent(context, node.path(Const.DATA)))
        }
      } catch (e: Exception) {
        Timber.v(e, "Failure while processing incoming data")
      }
    }

    private fun handleProtocolMessage(node: JsonNode) {
      model.pluginProtocol = getProtocolVersion(node)
      if (model.apiOutOfDate) {
        bus.post(MessageEvent(ProtocolEventType.PluginUpdateAvailable))
      }
      isHandshakeComplete = true
      bus.post(MessageEvent(ProtocolEventType.HandshakeComplete, true))
    }

    private fun getProtocolVersion(node: JsonNode): Int =
      try {
        Integer.parseInt(node.path(Const.DATA).asText())
      } catch (ignore: Exception) {
        2
      }
  }
